### PR TITLE
Feat/cache code lookups

### PR DESCRIPTION
**Pull request by 0xgregthedev**
This pr caches code lookups under the assumption that code is immutable.

### CODE REVIEW

File: crates/evm/core/src/fork/backend.rs [hunk 1]
```diff
@@ -603,12 +622,12 @@ impl SharedBackend {
         let (backend, backend_rx) = channel(1);
         let cache = Arc::new(FlushJsonBlockCacheDB(Arc::clone(db.cache())));
         let handler = BackendHandler::new(provider, db, backend_rx, pin_block);
-        (Self { backend, cache, data_accesses, chain, state_lookup }, handler)
+        (Self { backend, cache, data_accesses, chain, state_lookup, code_cache }, handler)
     }
 
     /// Updates the pinned block to fetch data from
-    pub fn set_pinned_block(&self, block: impl Into<BlockId>) -> eyre::Result<()> {
-        let req = BackendRequest::SetPinnedBlock(block.into());
```
Comment (odyslam) [inline crates/evm/core/src/fork/backend.rs:611]: Why are we making this change?

File: crates/evm/core/src/backend/code_cache.rs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+struct CodeDetected {
+    /// The block number at which the code was detected.
+    block_number: u64,
+    /// The code at that block.
+    code: Bytes,
+}
+
+#[derive(Debug, Default, Clone)]
+struct CodeHistory {
```
Comment (odyslam) [inline crates/evm/core/src/backend/code_cache.rs:16]: Why do we need the history?

**Review by odyslam — APPROVED** [review]
LGTM

I have a few comments to better understand the code, but it looks good.

File: crates/evm/core/src/backend/code_cache.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+        self.cache_code(address, chain, block_number, code.clone());
+
+        Ok(code)
+    }
+
+    /// Check the cache for the code of an account at a specific block.
+    /// Returns the code if it is in the cache, otherwise None.
+    ///
+    /// If the account had code at the time of the block or earlier, it had code at the time of the
```
Comment (Qwuke) [inline crates/evm/core/src/backend/code_cache.rs:56]: This comment is pretty hard to comprehend, and the comments on the CodeHistory struct also don't describe why they are Option and what the states might mean. Can you expand on them?

File: crates/evm/core/src/fork/backend.rs (hunk 1 above)
Comment (0xgregthedev) [reply crates/evm/core/src/fork/backend.rs:611]: A BlockId can be pinned to a tag or a block number. 

The BackendHandler needs to be pinned to a specific block number to use the code cache.

If we used the BlockId with the latest tag in place of a block number then we wouldn't be tracking block numbers properly and instead would be overwriting one entry as the chain progresses.

File: crates/evm/core/src/backend/code_cache.rs (hunk 2 above)
Comment (0xgregthedev) [reply crates/evm/core/src/backend/code_cache.rs]: I updated the comments and types to be more clear, but it's the metadata for a cache entry, needed to track when the earliest we detected code was, and when the latest we didn't detect code was. 
If you are lte the no_code_detected_block_number then the address didn't have code. 
if you are gte the code_detected.block_number then return the immutable code.
If there is a cache entry and the requested block is between this range then you must fetch a new result.

File: crates/evm/core/src/backend/code_cache.rs (hunk 3 above)
Comment (0xgregthedev) [reply crates/evm/core/src/backend/code_cache.rs:59]: Updated type defs and comments, lmk if still confusing.

**Review by odyslam — APPROVED** [review]
Comments provided clarity. Exciting to see this finished

**Pull request merged**